src/zkevm_specs/evm/instruction.py

```diff
@@ -107,7 +108,24 @@ def constrain_bool(self, num: FQ):
     def constrain_gas_left_not_underflow(self, gas_left: FQ):
         self.range_check(gas_left, N_BYTES_GAS)
 
-    def constrain_step_state_transition(self, **kwargs: Mapping[str, Transition]):
+    def constrain_execution_state_transition(self):
+        curr, next = self.curr.execution_state, self.next.execution_state
+
+        # ExecutionState transition constraint for special ones
+        if curr == ExecutionState.EndTx:
+            assert next in [ExecutionState.BeginTx, ExecutionState.EndBlock]
+        elif curr == ExecutionState.EndBlock:
+            assert next == ExecutionState.EndBlock
+
+        # Negation ExecutionState transition constraint for rest ones
+        if next == ExecutionState.BeginTx:
+            assert curr == ExecutionState.EndTx
+        elif next == ExecutionState.EndTx:
+            assert curr.halts() or curr == ExecutionState.BeginTx
```

Comment:
I'm wondering if this should be:
```python
assert (curr.halts() and curr.is_root) or curr == ExecutionState.BeginTx)
```

For example, if we're in a call at depth 3 and we RETURN/STOP, `curr.halts()` is true, but the execution continues to the call at depth 2.

Reply:
The checks here is more like "all possible `ExecutionState` transiton" when we are defining the state transition diagram.

When one can transit to multiple, its constraint should make sure the transition is reasonable. For example, in #100, we constrain something like this ([reference](
https://github.com/appliedzkp/zkevm-specs/pull/100/files#diff-09a605ab107c1b57880e784cd75a70e238f1ebf114bd4bf1d535671bff929f14R15-R17)):

```python
# Go to EndTx only when is_root
is_to_end_tx = instruction.is_equal(instruction.next.execution_state, ExecutionState.EndTx)
instruction.constrain_equal(instruction.curr.is_root + is_to_end_tx, 2 * instruction.curr.is_root * is_to_end_tx)
```